src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -23,91 +23,179 @@
 using namespace realm;
 using namespace realm::_impl;
 
-void DeepChangeChecker::find_related_tables(std::vector<RelatedTable>& out, Table const& table)
+void DeepChangeChecker::find_all_related_tables(std::vector<RelatedTable>& out, Table const& table,
+                                                const std::vector<TableKey>& tables_in_filters)
 {
     auto table_key = table.get_key();
-    if (any_of(begin(out), end(out), [=](auto& tbl) {
+    // If the currently looked at `table` is already part of the `std::vector<RelatedTable>` (possibly
+    // due to another path involving it) we do not need to traverse further and can return.
+    if (any_of(begin(out), end(out), [&](const auto& tbl) {
             return tbl.table_key == table_key;
         }))
         return;
 
+    // If a filter is set and the table is not part of the filter, it can be skipped.
+    if (tables_in_filters.size() != 0) {
+        if (none_of(begin(tables_in_filters), end(tables_in_filters), [&](const auto& filtered_table_key) {
+                return filtered_table_key == table_key;
+            })) {
+            return;
+        }
+    }
+
     // We need to add this table to `out` before recurring so that the check
     // above works, but we can't store a pointer to the thing being populated
     // because the recursive calls may resize `out`, so instead look it up by
-    // index every time
+    // index every time.
     size_t out_index = out.size();
     out.push_back({table_key, {}});
 
     for (auto col_key : table.get_column_keys()) {
         auto type = table.get_column_type(col_key);
+        // If a column within the `table` does link to another table it needs to be added to `table`'s
+        // links.
         if (type == type_Link || type == type_LinkList) {
             out[out_index].links.push_back({col_key.value, type == type_LinkList});
-            find_related_tables(out, *table.get_link_target(col_key));
+            // Finally this function needs to be called again to traverse all linked tables using the
+            // just found link.
+            find_all_related_tables(out, *table.get_link_target(col_key), tables_in_filters);
         }
     }
+    if (tables_in_filters.size() != 0) {
+        // Backlinks can only come into consideration when added via key paths.
+        // If there are not `tables_in_filter` we can skip this part.
+        table.for_each_backlink_column([&](ColKey column_key) {
+            auto target_table = table.get_link_target(column_key);
+            auto target_table_key = target_table->get_key();
+            if (any_of(tables_in_filters.begin(), tables_in_filters.end(), [target_table_key](auto table_key) {
+                    return target_table_key == table_key;
+                })) {
```

Comment:
This formatting's pretty awkward to read and would benefit from a local variable for the result of any_of rather than a multi-line conditional inside the if.